File: src/lib/ekhata/nepaliSentenceIntelligence.ts
/**
 * Nepali Sentence Intelligence — meaning, context, and corrupted-text repair.
 *
 * Handles:
 * - OCR / encoding corruption in Devanagari (replacement chars, missing matras)
 * - Clause-level structure (agent/recipient/source via postpositions)
 * - Sentence meaning synthesis from context — not literal token match
 *
 * Self-contained: no API, no downloads.
 */

/** Characters that indicate OCR/encoding corruption */
const CORRUPTION_RE =
  /[\uFFFD\uFFFE\uFFFF\u25A1\u25A0\u25AB\u25AA\u25FB\u25FC\u2610\u2611\u2612□▯▢■◻◼⬜⬛\uE000-\uF8FF]/gu;

const DEVANAGARI_RE = /[\u0900-\u097F]/;

/** Strip matras, halant, nukta — keep consonant skeleton for fuzzy match */
function consonantSkeleton(word: string): string {
  return word
    .replace(/[\u093E-\u094F\u0962\u0963]/g, "")
    .replace(/[\u094D\u0903\u0902]/g, "")
    .replace(/[^\u0900-\u097F]/g, "");
}

function levenshtein(a: string, b: string): number {
  const m = a.length;
  const n = b.length;
  if (!m) return n;
  if (!n) return m;
  const dp = Array.from({ length: m + 1 }, () => new Array<number>(n + 1).fill(0));
  for (let i = 0; i <= m; i++) dp[i][0] = i;
  for (let j = 0; j <= n; j++) dp[0][j] = j;
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      dp[i][j] =
        a[i - 1] === b[j - 1]
          ? dp[i - 1][j - 1]
          : 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
    }
  }
  return dp[m][n];
}

/** Domain lexicon — legal, accounting, khata vocabulary */
const DOMAIN_LEXICON: string[] = [
  // Legal / regulatory
  "नियमवाली",
  "ऐन",
  "दफा",
  "नियम",
  "संशोधन",
  "प्रारम्भ",
  "संक्षिप्त",
  "नाम",
  "राजपत्र",
  "मिति",
  "प्रकाशित",
  "सरकारी",
  "गजेट",
  "अधिकार",
  "प्रावधान",
  "परिभाषा",
  "दण्ड",
  "जरिवाना",
  // Tax / accounting
  "आयकर",
  "कर",
  "भ्याट",
  "मूल्य",
  "अभिवृद्धि",
  "कर",
  "उद्योग",
  "पेट्रोलियम",
  "आय",
  "खर्च",
  "लाभ",
  "हानि",
  "खाता",
  "जम्मा",
  "बाँकी",
  "उधार",
  "तिर्नु",
  "बिक्री",
  "खरिद",
  "भुक्तानी",
  "प्राप्त",
  "देय",
  "सम्पत्ति",
  "दायित्व",
  "पुँजी",
  "नाफा",
  "नोक्सान",
  "तलब",
  "भाडा",
  "ब्याज",
  "ऋण",
  "नगद",
  "बैंक",
  // Common khata chat
  "राम",
  "श्याम",
  "ग्राहक",
  "आपूर्तिकर्ता",
  "बिल",
  "चलान",
  "इनभ्वाइस",
  "पान",
  "दर्ता",
  "आर्थिक",
  "वर्ष",
  "महिना",
];

/** Context templates: [regex with capture groups, replacement fn] */
const CONTEXT_REPAIRS: Array<{
  pattern: RegExp;
  repair: (match: RegExpMatchArray) => string;
}> = [
  {
    // पेट्रोलियम उद्योग (आयकर) [corrupt] , २०४१ → नियमवाली
    pattern:
      /(पेट्रोलियम\s+उद्योग\s*\(\s*आयकर\s*\))\s*[^\u0900-\u097F\u0964।,]{0,12}[\u0900-\u097F]{0,12}\s*[,，]\s*(२०\d{2})/u,
    repair: (m) => `${m[1]} नियमवाली, ${m[2]}`,
  },
  {
    // नेपाल [corrupt]पत्रमा प्रकाशित → नेपाल राजपत्रमा प्रकाशित
    pattern: /(नेपाल)\s+(?:[^\u0900-\u097F\u0964।]*[\u0900-\u097F]{0,4})?पत्र(मा\s+प्रकाशित)/u,
    repair: (m) => `${m[1]} राजपत्र${m[2]}`,
  },
  {
    // प्रकाशित [corrupt]ति : → प्रकाशित मिति :
    pattern: /(प्रकाशित)\s+(?:[^\u0900-\u097F\u0964।]*[\u0900-\u097F]{0,4})?ति\s*[:：]/u,
    repair: (m) => `${m[1]} मिति :`,
  },
  {
    // संक्षिप्त [corrupt] र प्रारम्भ → संक्षिप्त नाम र प्रारम्भ
    pattern: /(संक्षिप्त)\s+[^\u0900-\u097F\u0964।]{0,6}[\u0900-\u097F]{0,8}\s+(र\s+प्रारम्भ)/u,
    repair: (m) => `${m[1]} नाम ${m[2]}`,
  },
  {
    // आयकर ऐन, २०३१ को दफा [num] → preserve
    pattern: /(आयकर\s+ऐन[^।]*दफा\s+\d+)/u,
    repair: (m) => m[1],
  },
  {
    // [corrupt] सरकारी → सरकारी (leading corruption before सरकारी)
    pattern: /[^\u0900-\u097F\u0964।]{1,6}(सरकारी)/u,
    repair: (m) => m[1],
  },
];

export type PostpositionRole = "agent" | "recipient" | "source" | "location" | "possessive" | "purpose";

export interface ClauseAnalysis {
  text: string;
  agent: string | null;
  recipient: string | null;
  source: string | null;
  verb: string | null;
  object: string | null;
  postpositionRoles: Array<{ phrase: string; role: PostpositionRole }>;
  isQuestion: boolean;
  isNegated: boolean;
  domainHint: "legal" | "accounting" | "transaction" | "general";
}

export interface SentenceMeaning {
  originalText: string;
  repairedText: string;
  corruptionScore: number;
  clauses: ClauseAnalysis[];
  primaryIntent: string | null;
  summaryNepali: string;
  summaryEnglish: string;
}

function corruptionScore(text: string): number {
  const matches = text.match(CORRUPTION_RE);
  if (!matches) return 0;
  return Math.min(1, matches.length / Math.max(text.length / 20, 1));
}

function fuzzyLexiconRepair(token: string): string {
  const clean = token.replace(CORRUPTION_RE, "").trim();
  if (!clean || !DEVANAGARI_RE.test(clean)) return token;

  const skel = consonantSkeleton(clean);
  if (!skel || skel.length < 2) return token;

  let best = clean;
  let bestDist = Infinity;

  for (const lex of DOMAIN_LEXICON) {
    const lexSkel = consonantSkeleton(lex);
    if (!lexSkel) continue;
    const dist = levenshtein(skel, lexSkel);
    const threshold = Math.max(2, Math.floor(lexSkel.length * 0.45));
    if (dist <= threshold && dist < bestDist) {
      bestDist = dist;
      best = lex;
    }
  }

  return bestDist < Infinity ? best : token;
}

/** Repair corrupted Devanagari using context templates + lexicon fuzzy match */
export function repairCorruptedDevanagari(text: string): string {
  if (!text?.trim()) return "";

  let out = text.normalize("NFKC");

  // Remove corruption placeholder chars
  out = out.replace(CORRUPTION_RE, "");

  // Apply context-aware template repairs (legal/regulatory patterns)
  for (const { pattern, repair } of CONTEXT_REPAIRS) {
    out = out.replace(pattern, (...args) => repair(args as unknown as RegExpMatchArray));
  }

  // Word-level fuzzy repair for remaining Devanagari tokens
  const parts = out.split(/(\s+|[।.،,;:!?]+)/);
  out = parts
    .map((part) => {
      if (!DEVANAGARI_RE.test(part) || part.trim().length < 3) return part;
      // Skip dates like २०४१।१२।९
      if (/^[२०-९।.\-/]+$/u.test(part.trim())) return part;
      return fuzzyLexiconRepair(part);
    })
    .join("");

  return out.replace(/\s+/g, " ").trim();
}

const QUESTION_RE =
  /\b(k\s*ho|ke\s*ho|kasari|kina|kati|kaha|kun|what|how|why|when|where|who|\?|के\s*हो|कति|कसरी|किन|कहाँ)\b/i;
const NEGATION_RE =
  /\b(xaina|chaina|chhaina|bhayena|vayena|hudaina|hoina|gardina|not|never|no)\b/i;

const LEGAL_MARKERS =
  /\b(नियमवाली|ऐन|दफा|राजपत्र|संशोधन|प्रावधान|गजेट|law|act|section|regulation|rule)\b/i;
const ACCOUNTING_MARKERS =
  /\b(आयकर|भ्याट|कर|खाता|जम्मा|उधार|बिक्री|खरिद|तलब|भाडा|ब्याज|ऋण|debit|credit|vat|tds|salary|expense|sale|purchase|payment)\b/i;
const TRANSACTION_MARKERS =
  /\b(\d+|saya|hajar|lakh|tiryo|kinyo|becheko|diye|aayo|kharcha|udhaar|nagad|cash|sold|bought|paid|received)\b/i;

function detectDomain(clause: string): ClauseAnalysis["domainHint"] {
  if (LEGAL_MARKERS.test(clause)) return "legal";
  if (TRANSACTION_MARKERS.test(clause)) return "transaction";
  if (ACCOUNTING_MARKERS.test(clause)) return "accounting";
  return "general";
}

function extractPostpositionRoles(clause: string): ClauseAnalysis["postpositionRoles"] {
  const roles: ClauseAnalysis["postpositionRoles"] = [];
  const soft = clause.replace(/[^\w\s\u0900-\u097F.]/g, " ").replace(/\s+/g, " ").trim();

  const patterns: Array<[RegExp, PostpositionRole]> = [
    [/([^\s]+(?:\s+[^\s]+)?)\s+(?:le|ले)\b/gi, "agent"],
    [/([^\s]+(?:\s+[^\s]+)?)\s+(?:lai|लाई)\b/gi, "recipient"],
    [/([^\s]+(?:\s+[^\s]+)?)\s+(?:bata|बाट)\b/gi, "source"],
    [/([^\s]+(?:\s+[^\s]+)?)\s+(?:ma|मा)\b/gi, "location"],
    [/([^\s]+(?:\s+[^\s]+)?)\s+(?:ko|को)\b/gi, "possessive"],
    [/([^\s]+(?:\s+[^\s]+)?)\s+(?:ko\s+lagi|को\s+लागि)\b/gi, "purpose"],
  ];

  for (const [re, role] of patterns) {
    for (const m of soft.matchAll(re)) {
      const phrase = (m[1] ?? "").trim();
      if (phrase && phrase.length >= 2) roles.push({ phrase, role });
    }
  }

  return roles;
}

function extractVerb(clause: string): string | null {
  const verbs =
    /\b(tiryo|tireko|kinyo|kinye|kineko|becheko|beche|bikyo|diye|diyo|aayo|aayeko|kharcha|gareko|garyo|प्रकाशित|संशोधन|प्रारम्भ|sold|bought|paid|received|expense)\b/i;
  const m = clause.match(verbs);
  return m ? m[1].toLowerCase() : null;
}

/** Split text into clauses on Nepali/English discourse boundaries */
export function segmentClauses(text: string): string[] {
  const normalized = text.replace(/\s+/g, " ").trim();
  if (!normalized) return [];

  const raw = normalized.split(/(?<=[।.;])\s+|(?:\s+)(?:ra|र|tara|तर|bhane|भने|ki|कि|so|therefore|ani|अनि)\s+/i);
  return raw.map((c) => c.trim()).filter((c) => c.length >= 3);
}

/** Analyze one clause for grammatical roles and domain */
export function analyzeClause(clause: string): ClauseAnalysis {
  const postpositionRoles = extractPostpositionRoles(clause);
  let agent: string | null = null;
  let recipient: string | null = null;
  let source: string | null = null;

  for (const { phrase, role } of postpositionRoles) {
    if (role === "agent" && !agent) agent = phrase;
    if (role === "recipient" && !recipient) recipient = phrase;
    if (role === "source" && !source) source = phrase;
  }

  // maile/timile/hamile
  if (/\bmaile\b/i.test(clause)) agent = "Self";
  if (/\btimile\b/i.test(clause)) agent = "Timi";
  if (/\bhamile\b/i.test(clause)) agent = "Hami";

  const verb = extractVerb(clause);

  // Object: noun before verb or after "ko"
  let object: string | null = null;
  const objMatch = clause.match(
    /\b([a-zA-Z\u0900-\u097F]{2,25})\s+(?:kinyo|kinye|kineko|becheko|beche|bikyo|tiryo|diye)\b/i,
  );
  if (objMatch) object = objMatch[1];

  return {
    text: clause,
    agent,
    recipient,
    source,
    verb,
    object,
    postpositionRoles,
    isQuestion: QUESTION_RE.test(clause),
    isNegated: NEGATION_RE.test(clause),
    domainHint: detectDomain(clause),
  };
}

function buildSummary(clauses: ClauseAnalysis[], repaired: string): { ne: string; en: string } {
  if (!clauses.length) {
    return { ne: repaired.slice(0, 120), en: repaired.slice(0, 120) };
  }

  const main = clauses[0];
  const parts: string[] = [];
  const enParts: string[] = [];

  if (main.domainHint === "legal") {
    parts.push("कानूनी/नियामक पाठ");
    enParts.push("Legal/regulatory text");
  } else if (main.domainHint === "transaction") {
    parts.push("लेनदेन सम्बन्धी वाक्य");
    enParts.push("Transaction-related sentence");
  } else if (main.domainHint === "accounting") {
    parts.push("लेखा/कर सम्बन्धी वाक्य");
    enParts.push("Accounting/tax-related sentence");
  }

  if (main.agent) {
    parts.push(`कर्ता: ${main.agent}`);
    enParts.push(`Agent: ${main.agent}`);
  }
  if (main.recipient) {
    parts.push(`प्राप्तकर्ता: ${main.recipient}`);
    enParts.push(`Recipient: ${main.recipient}`);
  }
  if (main.source) {
    parts.push(`स्रोत: ${main.source}`);
    enParts.push(`Source: ${main.source}`);
  }
  if (main.verb) {
    parts.push(`क्रिया: ${main.verb}`);
    enParts.push(`Verb: ${main.verb}`);
  }
  if (main.isQuestion) {
    parts.push("प्रश्न");
    enParts.push("Question");
  }
  if (main.isNegated) {
    parts.push("नकारात्मक");
    enParts.push("Negated");
  }

  return {
    ne: parts.join(" · ") || repaired.slice(0, 100),
    en: enParts.join(" · ") || repaired.slice(0, 100),
  };
}

function inferPrimaryIntent(clauses: ClauseAnalysis[]): string | null {
  for (const c of clauses) {
    if (c.isQuestion) return "question";
    if (c.domainHint === "legal") return "legal_reference";
    if (c.isNegated) return "negation";
    if (c.recipient && /\b(diye|diyo|becheko|udhaar)\b/i.test(c.text)) return "credit_sale";
    if (c.agent && /\b(tiryo|tireko|aayo)\b/i.test(c.text)) return "payment_received";
    if (c.recipient && /\b(tiryo|tireko)\b/i.test(c.text)) return "payment_made";
    if (c.source && /\b(kineko|kharid|kin)\b/i.test(c.text)) return "purchase";
    if (/\b(becheko|bikri|sold)\b/i.test(c.text)) return "sale";
    if (/\b(kharcha|expense)\b/i.test(c.text)) return "expense";
  }
  return null;
}

/** Full sentence meaning analysis with optional corruption repair */
export function analyzeSentenceMeaning(rawText: string): SentenceMeaning {
  const originalText = (rawText || "").trim();
  const hasDevanagari = DEVANAGARI_RE.test(originalText);
  const hasCorruption = CORRUPTION_RE.test(originalText);

  const repairedText =
    hasDevanagari && (hasCorruption || corruptionScore(originalText) > 0)
      ? repairCorruptedDevanagari(originalText)
      : originalText;

  const workText = repairedText || originalText;
  const clauses = segmentClauses(workText).map(analyzeClause);
  const { ne, en } = buildSummary(clauses, workText);

  return {
    originalText,
    repairedText: workText,
    corruptionScore: corruptionScore(originalText),
    clauses,
    primaryIntent: inferPrimaryIntent(clauses),
    summaryNepali: ne,
    summaryEnglish: en,
  };
}

/** Compact context block for LLM / grammar brain — sentence-level meaning */
export function synthesizeSentenceContext(message: string, maxChars = 900): string {
  const analysis = analyzeSentenceMeaning(message);
  if (!analysis.clauses.length && analysis.corruptionScore === 0) return "";

  const lines = [
    "[NEPALI SENTENCE CONTEXT]",
    "Interpret meaning from clause structure and postpositions — not corrupted glyphs.",
  ];

  if (analysis.corruptionScore > 0) {
    lines.push(`OCR repair applied (corruption score: ${analysis.corruptionScore.toFixed(2)})`);
    if (analysis.repairedText !== analysis.originalText) {
      lines.push(`Repaired: ${analysis.repairedText.slice(0, 200)}`);
    }
  }

  if (analysis.primaryIntent) {
    lines.push(`Primary intent signal: ${analysis.primaryIntent}`);
  }

  for (const clause of analysis.clauses.slice(0, 3)) {
    const hints: string[] = [];
    if (clause.agent) hints.push(`agent(le)=${clause.agent}`);
    if (clause.recipient) hints.push(`recipient(lai)=${clause.recipient}`);
    if (clause.source) hints.push(`source(bata)=${clause.source}`);
    if (clause.verb) hints.push(`verb=${clause.verb}`);
    if (clause.domainHint !== "general") hints.push(`domain=${clause.domainHint}`);
    if (hints.length) {
      lines.push(`Clause: ${clause.text.slice(0, 80)} → ${hints.join(", ")}`);
    }
  }

  lines.push(`Meaning: ${analysis.summaryEnglish}`);

  const out = lines.join("\n").trim();
  return out.length > maxChars ? `${out.slice(0, maxChars - 3)}...` : out;
}

File: erp_bot/src/nlu/nepali_sentence_intelligence.py
"""Nepali sentence intelligence — OCR repair, clause context, meaning synthesis."""

from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass, field
from typing import Literal

CORRUPTION_RE = re.compile(
    r"[\uFFFD\uFFFE\uFFFF\u25A1\u25A0\u25AB\u25AA\u25FB\u25FC\u2610\u2611\u2612"
    r"□▯▢■◻◼⬜⬛\uE000-\uF8FF]"
)
DEVANAGARI_RE = re.compile(r"[\u0900-\u097F]")

DOMAIN_LEXICON = [
    "नियमवाली", "ऐन", "दफा", "नियम", "संशोधन", "प्रारम्भ", "संक्षिप्त", "नाम",
    "राजपत्र", "मिति", "प्रकाशित", "सरकारी", "आयकर", "कर", "भ्याट", "उद्योग",
    "पेट्रोलियम", "उधार", "बिक्री", "खरिद", "भुक्तानी", "सम्पत्ति", "दायित्व",
    "पुँजी", "नाफा", "तलब", "भाडा", "ब्याज", "ऋण", "नगद", "बैंक",
]

CONTEXT_REPAIRS: list[tuple[re.Pattern[str], str]] = [
    (
        re.compile(
            r"(पेट्रोलियम\s+उद्योग\s*\(\s*आयकर\s*\))\s*[^\u0900-\u097F\u0964।,]{0,12}"
            r"[\u0900-\u097F]{0,12}\s*[,，]\s*(२०\d{2})"
        ),
        r"\1 नियमवाली, \2",
    ),
    (
        re.compile(
            r"(नेपाल)\s+(?:[^\u0900-\u097F\u0964।]*[\u0900-\u097F]{0,4})?पत्र(मा\s+प्रकाशित)"
        ),
        r"\1 राजपत्र\2",
    ),
    (
        re.compile(
            r"(प्रकाशित)\s+(?:[^\u0900-\u097F\u0964।]*[\u0900-\u097F]{0,4})?ति\s*[:：]"
        ),
        r"\1 मिति :",
    ),
    (
        re.compile(r"(संक्षिप्त)\s+[^\u0900-\u097F\u0964।]{0,6}[\u0900-\u097F]{0,8}\s+(र\s+प्रारम्भ)"),
        r"\1 नाम \2",
    ),
]

PostpositionRole = Literal["agent", "recipient", "source", "location", "possessive", "purpose"]
DomainHint = Literal["legal", "accounting", "transaction", "general"]


def _consonant_skeleton(word: str) -> str:
    word = re.sub(r"[\u093E-\u094F\u0962\u0963\u094D\u0903\u0902]", "", word)
    return re.sub(r"[^\u0900-\u097F]", "", word)


def _levenshtein(a: str, b: str) -> int:
    if not a:
        return len(b)
    if not b:
        return len(a)
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a, 1):
        curr = [i]
        for j, cb in enumerate(b, 1):
            curr.append(min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + (ca != cb)))
        prev = curr
    return prev[-1]


def corruption_score(text: str) -> float:
    matches = CORRUPTION_RE.findall(text)
    if not matches:
        return 0.0
    return min(1.0, len(matches) / max(len(text) / 20, 1))


def repair_corrupted_devanagari(text: str) -> str:
    if not text or not text.strip():
        return ""
    out = unicodedata.normalize("NFKC", text.strip())
    out = CORRUPTION_RE.sub("", out)
    for pattern, repl in CONTEXT_REPAIRS:
        out = pattern.sub(repl, out)

    parts = re.split(r"(\s+|[।.،,;:!?]+)", out)
    repaired: list[str] = []
    for part in parts:
        if not DEVANAGARI_RE.search(part) or len(part.strip()) < 3:
            repaired.append(part)
            continue
        if re.fullmatch(r"[२०-९।.\-/]+", part.strip()):
            repaired.append(part)
            continue
        skel = _consonant_skeleton(CORRUPTION_RE.sub("", part))
        best, best_dist = part, 999
        for lex in DOMAIN_LEXICON:
            lex_skel = _consonant_skeleton(lex)
            if not lex_skel:
                continue
            dist = _levenshtein(skel, lex_skel)
            threshold = max(2, int(len(lex_skel) * 0.45))
            if dist <= threshold and dist < best_dist:
                best_dist = dist
                best = lex
        repaired.append(best if best_dist < 999 else part)
    out = "".join(repaired)
    return re.sub(r"\s+", " ", out).strip()


def segment_clauses(text: str) -> list[str]:
    normalized = re.sub(r"\s+", " ", (text or "").strip())
    if not normalized:
        return []
    raw = re.split(
        r"(?<=[।.;])\s+|(?:\s+)(?:ra|र|tara|तर|bhane|भने|ki|कि|so|therefore|ani|अनि)\s+",
        normalized,
        flags=re.I,
    )
    return [c.strip() for c in raw if c.strip() and len(c.strip()) >= 3]


@dataclass
class ClauseAnalysis:
    text: str
    agent: str | None = None
    recipient: str | None = None
    source: str | None = None
    verb: str | None = None
    object: str | None = None
    postposition_roles: list[tuple[str, PostpositionRole]] = field(default_factory=list)
    is_question: bool = False
    is_negated: bool = False
    domain_hint: DomainHint = "general"


@dataclass
class SentenceMeaning:
    original_text: str
    repaired_text: str
    corruption_score: float
    clauses: list[ClauseAnalysis]
    primary_intent: str | None
    summary_nepali: str
    summary_english: str


_QUESTION_RE = re.compile(
    r"\b(k\s*ho|ke\s*ho|kasari|kina|kati|what|how|why|when|where|who|\?|के\s*हो)\b",
    re.I,
)
_NEGATION_RE = re.compile(
    r"\b(xaina|chaina|chhaina|bhayena|hudaina|hoina|not|never|no)\b",
    re.I,
)


def analyze_clause(clause: str) -> ClauseAnalysis:
    soft = re.sub(r"[^\w\s\u0900-\u097F.]", " ", clause)
    soft = re.sub(r"\s+", " ", soft).strip()
    roles: list[tuple[str, PostpositionRole]] = []
    agent = recipient = source = None

    for pattern, role in [
        (r"([^\s]+(?:\s+[^\s]+)?)\s+(?:le|ले)\b", "agent"),
        (r"([^\s]+(?:\s+[^\s]+)?)\s+(?:lai|लाई)\b", "recipient"),
        (r"([^\s]+(?:\s+[^\s]+)?)\s+(?:bata|बाट)\b", "source"),
    ]:
        for m in re.finditer(pattern, soft, re.I):
            phrase = (m.group(1) or "").strip()
            if phrase:
                roles.append((phrase, role))
                if role == "agent" and not agent:
                    agent = phrase
                elif role == "recipient" and not recipient:
                    recipient = phrase
                elif role == "source" and not source:
                    source = phrase

    if re.search(r"\bmaile\b", soft, re.I):
        agent = "Self"

    verb_m = re.search(
        r"\b(tiryo|tireko|kinyo|kineko|becheko|diye|aayo|kharcha|gareko|प्रकाशित)\b",
        soft,
        re.I,
    )
    verb = verb_m.group(1).lower() if verb_m else None

    domain: DomainHint = "general"
    if re.search(r"नियमवाली|ऐन|दफा|राजपत्र|संशोधन", soft):
        domain = "legal"
    elif re.search(r"\d+|tiryo|kinyo|becheko|udhaar|nagad", soft, re.I):
        domain = "transaction"
    elif re.search(r"आयकर|भ्याट|कर|खाता|उधार|बिक्री", soft):
        domain = "accounting"

    return ClauseAnalysis(
        text=clause,
        agent=agent,
        recipient=recipient,
        source=source,
        verb=verb,
        postposition_roles=roles,
        is_question=bool(_QUESTION_RE.search(clause)),
        is_negated=bool(_NEGATION_RE.search(clause)),
        domain_hint=domain,
    )


def analyze_sentence_meaning(raw_text: str) -> SentenceMeaning:
    original = (raw_text or "").strip()
    has_dev = bool(DEVANAGARI_RE.search(original))
    has_corr = bool(CORRUPTION_RE.search(original))
    repaired = (
        repair_corrupted_devanagari(original)
        if has_dev and (has_corr or corruption_score(original) > 0)
        else original
    )
    work = repaired or original
    clauses = [analyze_clause(c) for c in segment_clauses(work)]

    summary_parts = []
    en_parts = []
    if clauses:
        main = clauses[0]
        if main.domain_hint == "legal":
            summary_parts.append("कानूनी/नियामक पाठ")
            en_parts.append("Legal/regulatory text")
        if main.agent:
            summary_parts.append(f"कर्ता: {main.agent}")
            en_parts.append(f"Agent: {main.agent}")
        if main.recipient:
            summary_parts.append(f"प्राप्तकर्ता: {main.recipient}")
            en_parts.append(f"Recipient: {main.recipient}")

    primary = None
    for c in clauses:
        if c.is_question:
            primary = "question"
            break
        if c.domain_hint == "legal":
            primary = "legal_reference"
            break
        if c.recipient and re.search(r"diye|becheko|udhaar", c.text, re.I):
            primary = "credit_sale"
            break
        if c.agent and re.search(r"tiryo|aayo", c.text, re.I):
            primary = "payment_received"
            break

    return SentenceMeaning(
        original_text=original,
        repaired_text=work,
        corruption_score=corruption_score(original),
        clauses=clauses,
        primary_intent=primary,
        summary_nepali=" · ".join(summary_parts) or work[:100],
        summary_english=" · ".join(en_parts) or work[:100],
    )


def synthesize_sentence_context(message: str, max_chars: int = 900) -> str:
    analysis = analyze_sentence_meaning(message)
    if not analysis.clauses and analysis.corruption_score == 0:
        return ""
    lines = [
        "[NEPALI SENTENCE CONTEXT]",
        "Interpret meaning from clause structure and postpositions — not corrupted glyphs.",
    ]
    if analysis.corruption_score > 0:
        lines.append(f"OCR repair applied (score: {analysis.corruption_score:.2f})")
        if analysis.repaired_text != analysis.original_text:
            lines.append(f"Repaired: {analysis.repaired_text[:200]}")
    if analysis.primary_intent:
        lines.append(f"Primary intent signal: {analysis.primary_intent}")
    for clause in analysis.clauses[:3]:
        hints = []
        if clause.agent:
            hints.append(f"agent(le)={clause.agent}")
        if clause.recipient:
            hints.append(f"recipient(lai)={clause.recipient}")
        if clause.verb:
            hints.append(f"verb={clause.verb}")
        if hints:
            lines.append(f"Clause: {clause.text[:80]} → {', '.join(hints)}")
    lines.append(f"Meaning: {analysis.summary_english}")
    out = "\n".join(lines).strip()
    return out if len(out) <= max_chars else out[: max_chars - 3] + "..."
